File: engine/lz77.py
window_size = 13      # Size of the sliding window
search_buffer = 7     # Size of the search buffer
look_ahead_buffer = 6 # Size of the look-ahead buffer

def compress_lz77(input_text):
    compressed_output = ""
    position = 0
    input_length = len(input_text)
    search_buffer = 13
    look_ahead_buffer = 6
    
    while position < input_length:
        token = {'offset': 0, 'length_of_match': 0}
        
        # Calculate the maximum offset based on the search buffer
        max_offset = min(position, search_buffer)
        
        # Calculate the maximum search length based on the look-ahead buffer
        max_search_length = min(input_length - position, look_ahead_buffer)
        
        # Search for the longest match in the search buffer
        for offset in range(1, max_offset + 1):
            length = 0
            while (length < max_search_length and
                   position - offset + length >= 0 and
                   input_text[position - offset + length] == input_text[position + length]):
                length += 1
            
            # Update the token if a longer match is found
            if length > token['length_of_match']:
                token['offset'] = offset
                token['length_of_match'] = length
        
        # Append the token to the compressed output
        if token['length_of_match'] > 2:
            compressed_output += f"<{token['offset']},{token['length_of_match']}>"
            position += token['length_of_match']
        else:
            compressed_output += input_text[position]
            position += 1
    
    return compressed_output


def decompress_lz77(compressed_text):
    output = []
    pos = 0
    while pos < len(compressed_text):
        if compressed_text[pos] == '<':
            temp = ''
            pos += 1
            while compressed_text[pos]!= ',':
                temp += compressed_text[pos]
                pos += 1
            offset = int(temp)
            pos += 1
            temp = ''
            while compressed_text[pos]!= '>':
                temp += compressed_text[pos]
                pos += 1
            length = int(temp)
            pos += 1
            start = len(output) - offset
            for i in range(length):
                output.append(output[start + i])
        else:
            output.append(compressed_text[pos])
            pos += 1
    return ''.join(output)

def run():
    input_text = input("Enter Input: ").strip()
    
    if not input_text:
        print("Empty input. Please provide a valid input string.")
        return
    
    tokens = compress_lz77(input_text)
    print("Compression:")
    print("<Offset, Length Of Match>: ")
    for token in tokens:
        print(f"<{token['offset']}, {token['length_of_match']}>")
    
    compressed_output = ''
    for i, token in enumerate(tokens):
        if token['offset'] == 0 and token['length_of_match'] == 0:
            compressed_output += input_text[i]
        else:
            compressed_output += f"<{token['offset']},{token['length_of_match']}>"
    
    print("Compressed Output: ", compressed_output)
    print("Decompressed Output: ", decompress_lz77(tokens, input_text))



if __name__ == "__main__":
    input_text = input("Enter Input: ").strip()
    compressed = compress_lz77(input_text)
    print(compressed)

    decompressed = decompress_lz77(compressed)
    print(decompressed)

File: lz77/lz77.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int window_size = 13;      // Size of the sliding window
int search_buffer = 7;     // Size of the search buffer
int look_ahead_buffer = 6; // Size of the look-ahead buffer

struct Token
{
    int offset;          // Offset of the longest match
    int length_of_match; // Length of the longest match
};

vector<Token> compression_lz77(const string& input)
{
    int inputLength = input.length();
    vector<Token> data;
    int position = 0;

    while (position < inputLength)
    {
        Token token;
        token.offset = 0;
        token.length_of_match = 0;
      
        // Calculate the maximum offset based on the search buffer
        int max_offset = (position - search_buffer) < 0 ? position : search_buffer;
      
        // Calculate the maximum search length based on the look-ahead buffer
        int max_search_length = (position + look_ahead_buffer) > inputLength ? inputLength - position : look_ahead_buffer;

        // Search for the longest match in the search buffer
        for (int offset = 1; offset <= max_offset; offset++)
        {
            int len = 0;
            while (len < max_search_length && input[position - offset + len] == input[position + len])
            {
                len++;
            }
          
            // Update the token if a longer match is found
            if (len > token.length_of_match)
            {
                token.offset = offset;
                token.length_of_match = len;
            }
        }

        data.push_back(token);
        position += token.length_of_match + 1;
    }

    return data;
}

string decompress_lz77(const vector<Token>& tokens, const string& original_input)
{
    string tmp;
    int pos = 0;
    int inputLength = original_input.length();

    for (const auto& token : tokens)
    {
        if (token.offset != 0)
        {
            int start = pos - token.offset;
            int len = token.length_of_match;
            // Copy the matched substring
            while (len > 0)
            {
                tmp += tmp[start];
                start++;
                len--;
                pos++;
            }
        }
        if (pos < inputLength) {
            tmp += original_input[pos];
            pos++;
        }
    }

    return tmp;
}

void run()
{
    string input;
    cout << "Enter Input: ";
    getline(cin, input);

    if (input.empty()) {
        cout << "Empty input. Please provide a valid input string." << endl;
        return;
    }

    auto tokens = compression_lz77(input);
    cout << "Compression:" << endl;
    cout << "<Offset, Length Of Match>: " << endl;
    for (const auto& token : tokens)
    {
        cout << "<" << token.offset << ", " << token.length_of_match << ">" << endl;
    }

    string compressed_output;
    for (size_t i = 0; i < tokens.size(); i++)
    {
        if (tokens[i].offset == 0 && tokens[i].length_of_match == 0)
        {
            compressed_output += input[i];
        }
        else
        {
            compressed_output += "<" + to_string(tokens[i].offset) + "," + to_string(tokens[i].length_of_match) + ">";
        }
    }
    
    cout << "Compressed Output: " << compressed_output << endl;
    cout << "Decompressed Output: " << decompress_lz77(tokens, input) << endl;
}

int main()
{
    run();
    return 0;
}
